assertj-error-prone/src/main/java/com/palantir/assertj/errorprone/AssertjSize.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.assertj.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Symbol;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+@AutoService(AssertjChecker.class)
+public final class AssertjSize implements AssertjChecker {
+
+    private static final String DESCRIPTION =
+            "Prefer AsssertJ size asserts for more debugging information than simple integer comparisons.";
+
+    private static final Matcher<ExpressionTree> sizeMatcher = Matchers.ignoreParens(Matchers.anyOf(
+            MethodMatchers.instanceMethod()
+                    .onDescendantOf(CharSequence.class.getName())
+                    .named("length")
+                    .withParameters(),
+            MethodMatchers.instanceMethod()
+                    .onDescendantOf(Map.class.getName())
```

Comment:
Refactoring assertThat(map.size) to assertThat(map) may fail if the map type also implements iterable due to assertThat(Map) and assertThat(Iterable) ambiguity.